### ISSUE TITLE
Add embedding and feature-engineering layer for GBLM inputs

### ISSUE THREAD

**Issue by kyo219** [kyo219/Gradient-Boosting-Language-Model#1]
## Summary

Right now, the GBLM model is trained only on raw token IDs (previous `L` tokens → next token).
I’d like to add a **pluggable feature-engineering layer** so that we can flexibly attach additional features (correlations between tokens, stats over the context, embedding-based features, etc.) on top of the current setup.

In addition, I’d like to **integrate strong pre-trained word embeddings** (e.g. GloVe / fastText / word2vec / any publicly available “good” embedding) aligned to this corpus-specific vocab, and then build **embedding-based features** on top of that.

This issue is about:

- documenting the current feature representation, and  
- proposing a structure for:
  - where/how to plug in extra features, and
  - how to attach external embeddings and use them for feature engineering.

## Current feature representation

Currently, the training data coming out of `make_gblm_training_data()` looks like this:

- `X_ids`: shape `(N, L)`  
  → each row is **just the previous `L` token IDs in the context**, in order.
- `y`: shape `(N,)`  
  → each entry is **the next token ID** to be predicted.

So the model is trained on:

> “Given the last `L` token IDs, predict the next token ID.”

### Example (`context_length = L = 4`)

For illustration:

| sample | x_0    | x_1    | x_2 | x_3 | y  |
|--------|--------|--------|-----|-----|----|
| 0      | `<PAD>` | `<PAD>` | 5   | 10  | 7  |
| 1      | `<PAD>` | 5      | 10  | 7   | 3  |
| 2      | 5      | 10     | 7   | 3   | 9  |

Here:

- `5`, `10`, `7`, `9` are **token IDs in the vocab**, nothing more.
- They are effectively **consecutive integer labels (dummy codes)** for words.

LightGBM sees each column `x_j` as a **categorical feature** and learns splits like:

- `x_3 == 7`
- `x_2 ∈ {5, 10, 23}`, etc.

There is **no embedding or semantic feature yet** – the model only uses:

- “which token ID appears at which position in the last `L` tokens”

to predict the next token ID.

## Proposal: add a pluggable feature-engineering layer

Conceptually, I’d like to introduce a dedicated place for feature engineering:

- Keep `gblm_data` as-is (responsible only for `(context_ids, y)` generation).
- Add a **feature-building layer** in `gblm_model`, e.g.:

  - `gblm_model/features.py`
    - `FeatureConfig` dataclass to toggle features on/off
    - `build_features_for_lightgbm(X_ids, tokenizer, feature_cfg, embedding_matrix=None) -> (X, categorical_feature_indices)`

- Wire this into:
  - `gblm_model/train.py`:
    - call `build_features_for_lightgbm()` before constructing `lgb.Dataset`
  - `gblm_model/inference.py`:
    - reuse the same logic to build features for a single context at inference time

Rough sketch:

```python
# gblm_model/features.py
def build_features_for_lightgbm(
    X_ids: np.ndarray,         # (N, L) context token IDs
    tokenizer: Tokenizer,
    cfg: FeatureConfig,
    embedding_matrix: np.ndarray | None = None,  # (V, D) if provided
) -> tuple[np.ndarray, list[int]]:
    """
    Transform raw context token IDs into final features for LightGBM.
    - Left part: categorical features (token IDs per position)
    - Right part: engineered numeric features (stats, correlations, embeddings, etc.)
    """
    ...
    return X, categorical_feature_indices
```

Plan: attach pre-trained embeddings to this corpus

I’d like to use strong off-the-shelf embeddings (e.g. GloVe / fastText / word2vec / any “good” public embedding) and align them to the GBLM vocab.

High-level plan:
	1.	Load a pre-trained embedding file
	•	Something like:
	•	pretrained.vec / glove.txt / etc.
	•	Build a dictionary:
```python
emb_dict: dict[str, np.ndarray]  # word -> vector (dim = D)
```
	2.	Align with the tokenizer vocab
	•	For the GBLM tokenizer:
```python
V = len(tokenizer.itos)
embedding_matrix = np.zeros((V, D), dtype=np.float32)

for idx, token in enumerate(tokenizer.itos):
    if token in emb_dict:
        embedding_matrix[idx] = emb_dict[token]
    else:
        embedding_matrix[idx] = unk_vector  # e.g. zero, random, or a shared <UNK> vector
```
	•	This produces:
	•	embedding_matrix[id] -> embedding for that token
	•	Save it as something like:
	•	artifacts/embedding_matrix.npy

	3.	Use this embedding_matrix inside the feature-engineering layer
	•	build_features_for_lightgbm(...) will accept embedding_matrix and construct embedding-based features for each sample.

Examples of possible features

These don’t have to be implemented all at once, but they are use cases that this layer should make easy.

1. Context-level statistics

Simple scalars per sample, based only on token IDs:
	•	Number of non-<PAD> tokens (effective context length)
	•	Number of <UNK> tokens in the context
	•	Type-token ratio within the context
(unique tokens / total non-<PAD> tokens)

These are cheap and easy to add as numeric columns.

2. “Correlation-like” features between positions

For example:
	•	Co-occurrence-based features between positions:
	•	e.g. something like “how often token at position i and token at position j occur together in the corpus”
	•	Aggregated measures over the context:
	•	counts of particular token types, POS-like categories (if we ever tag them), etc.

The idea is to expose structure across positions in the context, not only “raw ID at each slot”.

3. Embedding-based features (using pre-trained embeddings)

Once we have a embedding_matrix (V, D) aligned to the vocab, we can add embedding-based features. For example:

3.1 Context embedding (mean-pooling)
For each sample:
	•	Map X_ids (shape (L,)) → embeddings (shape (L, D)).
	•	Mask out <PAD> tokens.
	•	Compute a mean-pooled context embedding:

```python
# X_ids: (N, L)
# embedding_matrix: (V, D)

emb_seq = embedding_matrix[X_ids]        # (N, L, D)
mask = (X_ids != tokenizer.pad_id)[..., None].astype(np.float32)  # (N, L, 1)
emb_sum = (emb_seq * mask).sum(axis=1)   # (N, D)
lengths = mask.sum(axis=1).clip(min=1.0) # (N, 1)
context_emb = emb_sum / lengths          # (N, D)
```

Then:
	•	Option A: concatenate context_emb directly as numeric features:
	•	Final X: [ token_id features | context_emb ]
	•	categorical_features = [0, 1, ..., L-1] (left part), right part numeric.
	•	Option B: optionally reduce context_emb dimensionality (e.g. PCA → D') before concatenating, if we want fewer numeric columns.

3.2 “Prototype” / cluster-based cosine similarities
To avoid adding a full V-dimensional vector per sample, we can use prototypes:
	•	Cluster the vocab embeddings into K clusters (e.g. k-means).
	•	Take the centroid embedding of each cluster.
	•	For each context:
	•	Compute cosine similarity between context_emb and each centroid (K values).
	•	Use these K cosine values as numeric features.

This gives us features like:
	•	“How similar is this context to prototype #k?”
(which could roughly correspond to a semantic cluster: e.g. “characters”, “places”, etc., depending on embeddings and corpus)

We could also:
	•	Define special hand-picked sets of words (e.g. “pronouns”, “verbs”, “location words”) and compute:
	•	max / mean cosine similarity between context_emb and embeddings of words in each set.

This keeps the dimensionality manageable, while still leveraging strong pre-trained embeddings.

Scope of this issue
	•	Document the current feature representation (done above).
	•	Introduce a FeatureConfig + build_features_for_lightgbm()-style hook in gblm_model.
	•	Introduce a way to load and align pre-trained embeddings to the tokenizer vocab:
	•	e.g. a helper script like scripts/build_embedding_matrix.py that:
	•	loads a pre-trained embedding file,
	•	builds embedding_matrix (V, D) aligned with tokenizer.itos,
	•	saves it under artifacts/embedding_matrix.npy.
	•	Wire the feature layer into:
	•	train.py (training pipeline)
	•	inference.py (generation pipeline)
	•	Add at least a minimal set of simple features as an initial example:
	•	context length, number of <UNK>, etc.
	•	Add at least one embedding-based feature as a baseline:
	•	e.g. mean-pooled context_emb (optionally reduced to lower dimension) concatenated as numeric features.
	•	Keep it easy to extend with more complex features:
	•	correlation-like features across positions,
	•	more advanced embedding-based features (prototype similarities, etc.),
	•	without changing the training/inference code paths.
